packages/hub/indexing/document-context.js

```diff
@@ -150,67 +178,123 @@ module.exports = class DocumentContext {
+        let pristineItem = await this._build(resource.type, resource.id, resource, includesTree, depth + 1);
+        assignMeta(pristineItem, jsonapiDoc.data[index]);
+
+        pristine.data.push(pristineItem);
+        rootItems.push(`${resource.type}/${resource.id}`);
+      }
+      assignMeta(pristine, jsonapiDoc);
+    } else {
+      let contentType = this.schema.types.get(type);
+      if (!contentType) {
+        log.warn("ignoring unknown document type=%s id=%s", type, id);
+        return;
+      }
+
+      if (depth === 0 && this.includePaths.length){
+        searchTree = {};
+        for (let segments of this.includePaths) {
+          this._buildSearchTree(searchTree, contentType, segments);
+        }
+      }
+
+      // we are going inside a parent document's includes, so we need
+      // our own type here.
+      if (depth > 0) {
+        searchDoc['type'] = type;
+      }
+      let userModel = new Model(contentType, jsonapiDoc, this.schema, this.read.bind(this));
+      await this._buildAttributes(contentType, jsonapiDoc, userModel, pristine, searchDoc);
+      await this._buildRelationships(contentType, jsonapiDoc, userModel, pristine, searchDoc, searchTree, depth);
+
+      assignMeta(pristine.data, jsonapiDoc);
     }
 
-    let userModel = new Model(contentType, jsonapiDoc, this.schema, this.read.bind(this));
-    await this._buildAttributes(contentType, jsonapiDoc, userModel, pristine, searchDoc);
-    await this._buildRelationships(contentType, jsonapiDoc, userModel, pristine, searchDoc, searchTree, depth);
+    if (depth === 0) {
+      this.pristineIncludes = this.pristineIncludes.filter(r => !rootItems.includes(`${r.type}/${r.id}`));
+    }
 
     // top level document embeds all the other pristine includes
     if (this.pristineIncludes.length > 0 && depth === 0) {
       pristine.included = uniqBy([pristine].concat(this.pristineIncludes), r => `${r.type}/${r.id}`).slice(1);
     }
 
-    // The next fields in the searchDoc get a "cardstack_" prefix so
-    // they aren't likely to collide with the user's attribute or
-    // relationship.
-
-    if (jsonapiDoc.meta) {
-      pristine.data.meta = Object.assign({}, jsonapiDoc.meta);
-    } else {
-      pristine.data.meta = {};
-    }
-
     if (depth > 0) {
       this.pristineIncludes.push(pristine.data);
     } else {
+      // The next fields in the searchDoc get a "cardstack_" prefix so
+      // they aren't likely to collide with the user's attribute or
+      // relationship.
       searchDoc.cardstack_pristine = pristine;
-      searchDoc.cardstack_references = this._references;
-      searchDoc.cardstack_realms = this.schema.authorizedReadRealms(type, jsonapiDoc);
-      authLog.trace("setting resource_realms for %s %s: %j", type, id, searchDoc.cardstack_realms);
+      if (!isCollection) {
+        searchDoc.cardstack_references = this._references;
+        searchDoc.cardstack_realms = this.schema.authorizedReadRealms(type, jsonapiDoc);
```

Comment:
I was about to say these aren't necessary anymore (because we have `realms` as a separate property on the document context). But maybe they are still used on included resources in the search index, as part of the process of filtering a response for only resources the user can see?

Reply:
yeah, those are definitely still used.

Reply:
or well I should amend that--we use that information, but I think it's in a different place after the pgsearch refactor

Reply:
let me delete it and just see what happens....

Reply:
yeah, it's not necessary. i'll clean this up a little more